docs/reference/configuration/configuration.md

```diff
@@ -1,48 +1,276 @@
+
+```
+Configuration parameters
+------------------------
+cellular.random_max_start_delay = 100 (macro name: "MBED_CONF_CELLULAR_RANDOM_MAX_START_DELAY")
+cellular.use-apn-lookup = 0 (macro name: "MBED_CONF_CELLULAR_USE_APN_LOOKUP")
+<output truncated for brevity>
+```
+
+The order in which overrides are considered is:
+
+ 1. Libraries override target configuration with `mbed_lib.json`.
+ 2. The application overrides target and library configuration with `mbed_app.json`
+
+#### Defining configuration parameters
+
+The configuration system understands configuration parameters that targets, libraries and applications define using a JSON object called "config".
+
+For example:
 
 ```JSON
 {
     "config": {
-        "main-stack-size": {
-            "value": 2000
-        }
+        "param1": {
+            "help": "The first configuration parameter",
+            "macro_name": "CUSTOM_MACRO_NAME",
+            "value": 0
+        },
+        "param2": {
+            "help": "The second configuration parameter",
+            "required": true
+        },
+        "param3": 10
     }
 }
 ```
 
-#### Configuring user spawned thread stacks
+You define a configuration parameter by specifying its name as the key and specifying its value either with a description object or by value. The JSON fragment above defines three configuration parameters named `param1`, `param2` and `param3`.
+
+Above, the configuration parameters `param1` and `param2` are defined using a description object. The description object supports the following keys:
+
+  - `help`: an optional help message that describes the purpose of the parameter.
+  - `value`: an optional field that defines the value of the parameter.
+  - `required`: an optional key that specifies whether the parameter must have a value before compiling the code (`false` by default). It's not possible to compile a source tree with one or more required parameters that don't have a value. Generally, setting `required` to true is only useful when `value` is not set.
+  - `macro_name`: an optional name for the macro defined at compile time for this configuration parameter. The configuration system automatically figures out the corresponding macro name for a configuration parameter, but the user can override this automatically computed name by specifying `macro_name`.
+
+You define a macro by value by using an integer or string instead of the description object, such as `param3` above. Defining a parameter by value is equivalent to defining it with a description object with the key `value` set to the value in place of the description object, the key `help` not set, the key `macro_name` not set and the key `required` set to `false`. Defining a parameter by value is equivalent to a configuration parameter defined with a description object with the key `value` set to the value in place of the description object, the key `help` unset, the key `macro_name` unset, and the key `required` set to `false`.
```

Comment:
This paragraph contains the same sentence essentially duplicated.

Reply:
That's probably a merge/rebase mistake.